presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1348,11 +1356,27 @@ else if (expression instanceof LongLiteral) {
                     }
 
                     // otherwise, just use the expression as is
+                    boolean orderByOnExpression = false;
                     if (orderByExpression == null) {
                         orderByExpression = expression;
+                        orderByOnExpression = true;
                     }
 
                     ExpressionAnalysis expressionAnalysis = analyzeExpression(orderByExpression, sourceScope);
+                    if (orderByOnExpression) {
+                        // Try to resolve within orderByScope. If it is possible, then we warn - standard behavior differs from legacy one.
+                        for (Expression columnExpression : expressionAnalysis.getColumnReferences()) {
+                            Optional<ResolvedField> resolvedField = orderByScope.tryResolveField(columnExpression);
```

Comment:
An `ORDER BY` expression may contain a subquery which in turn may contain `Identifier`-s _resolvable_ with `orderByScope` but which should actually be resolved using different scope (subquery's) and would resolve to something else (shadowing).

To properly recognize column reference's "scope of origin" here you need to leverage analysis done by `analyzeExpression(orderByExpression, sourceScope)` which already took care of everything scope-related. In fact, you might need `RelationId` from #7354